feat: rework `StrictSeries` to `RelSeries`

jjaassoonn opened the pull request:
I amend the definition of `StrictSeries` according to [this thread](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/StrictSeries.20!4.233858.2C.20!4.233852/near/356952195)

Now a `RelSeries r` of length `n` is $a_0 \stackrel{r}{\to} a_1 \stackrel{r}{\to} \dots \stackrel{r}{\to} a_n$.

In case `r` is transitive, then `p : RelSeries r` implies that the underlying function of `r` is strictly monotonic.

Then if $\alpha$ is a preordered set, an `LTSeries` is a `RelSeries (. < .)` and its Krull dimension is defined as the supremum of length of `LTSeries`.

For future: now `CompositionSeries` is `RelSeries jordan_holder_lattice.is_maximal` and rework the Jordan Holder file


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
```
```suggestion
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+
+/-!
+# Series of a relation
+If `r` is a relation on `α` then a relation series of length `n` is a series
```
```suggestion

If `r` is a relation on `α` then a relation series of length `n` is a series
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+For any type `α`, each term of `α` gives a relation series with the right most index to be 0.
+-/
+@[simps!] def singleton (a : α) : RelSeries r where
+length := 0
+toFun := fun _ => a
+step := fun i => Fin.elim0 i
```
```suggestion
  length := 0
  toFun := fun _ => a
  step := fun i => Fin.elim0 i
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+step := fun i => Fin.elim0 i
+
+instance [IsEmpty α] : IsEmpty (RelSeries r) where
+false := fun x ↦ IsEmpty.false (x 0)
```
```suggestion
  false := fun x ↦ IsEmpty.false (x 0)
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+false := fun x ↦ IsEmpty.false (x 0)
+
+instance [Inhabited α] : Inhabited (RelSeries r) where
+default := singleton r default
```
```suggestion
  default := singleton r default
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+-/
+@[simps!]
+def OfLE (x : RelSeries r) {s : Rel α α} (h : r ≤ s) : RelSeries s where
+length := x.length
+toFun := x
+step := fun _ => h _ _ <| x.step _
```
```suggestion
  length := x.length
  toFun := x
  step := fun _ => h _ _ <| x.step _
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+step := fun _ => h _ _ <| x.step _
+
+lemma ofLE_length (x : RelSeries r) {s : Rel α α} (h : r ≤ s) :
+  (x.OfLE h).length = x.length := rfl
```
```suggestion
    (x.OfLE h).length = x.length := rfl
```

kbuzzard replied:
```suggestion
    (x.OfLE h).length = x.length := rfl
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+namespace LTSeries
+
+instance : Preorder (LTSeries α) :=
+Preorder.lift fun x => x.length
```
```suggestion
  Preorder.lift fun x => x.length
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+{ coe := fun x => x.toFun }
+
+lemma le_def (x y : LTSeries α) : x ≤ y ↔ x.length ≤ y.length :=
+Iff.rfl
```
```suggestion
  Iff.rfl
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+Iff.rfl
+
+lemma lt_def (x y : LTSeries α) : x < y ↔ x.length < y.length :=
+Iff.rfl
```
```suggestion
  Iff.rfl
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+Iff.rfl
+
+lemma top_len_unique [OrderTop (LTSeries α)] (p : LTSeries α) (hp : IsTop p) :
+  p.length = (⊤ : LTSeries α).length :=
+le_antisymm (@le_top (LTSeries α) _ _ _) (hp ⊤)
```
```suggestion
    p.length = (⊤ : LTSeries α).length :=
  le_antisymm (@le_top (LTSeries α) _ _ _) (hp ⊤)
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+le_antisymm (@le_top (LTSeries α) _ _ _) (hp ⊤)
+
+lemma top_len_unique' (H1 H2 : OrderTop (LTSeries α)) : H1.top.length = H2.top.length :=
+le_antisymm (H2.le_top H1.top) (H1.le_top H2.top)
```
```suggestion
  le_antisymm (H2.le_top H1.top) (H1.le_top H2.top)
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+le_antisymm (H2.le_top H1.top) (H1.le_top H2.top)
+
+lemma StrictMono (x : LTSeries α) : StrictMono x :=
+fun _ _ h => x.toRelSeries.StrictMono h
```
```suggestion
  fun _ _ h => x.toRelSeries.StrictMono h
```

kbuzzard replied:
```suggestion
  fun _ _ h => x.toRelSeries.StrictMono h
```

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,157 @@
+variable {β : Type _} [PartialOrder β]
+
+lemma Monotone (x : LTSeries β) : Monotone x :=
+fun _ _ h => by
```
```suggestion
  fun _ _ h => by
```

kbuzzard reviewed (commented):
Suggesting (my understanding of) Lean 4 whitespace conventions. But in general this looks fine to me.

semorrison commented:
Would you mind either adding the lemma(s) relating it to `List.Chain' r (List.ofFn s.toFun)`, or at least a comment about `Chain'`, so that users of this API are aware of the connection?

(If you really want to be thorough you could mention the `Quiver.path` connection too, but I think that is less important.)

jjaassoonn commented:
there is a very weird bug (I think this might be a bug), in `RelSeries.lean`, if I import `Mathlib.Data.List.ofFn`, then build will fail, but if I import something that imported `Mathlib.Data.List.ofFn` then it will build on github

If I import `Mathlib.Data.List.ofFn` directly and run `lake build`, then this is the (important bit?) of the error message:
```
./././Mathlib/Order/KrullDimension.lean:7:0: error: import Mathlib.Data.List.OfFn failed, environment already contains '_private.Mathlib.Data.List.OfFn.0.List.replicate.match_1.splitter' from Mathlib.Data.List.ofFn
```

one can see the failed run at [here](https://github.com/leanprover-community/mathlib4/actions/runs/5114761577/jobs/9195334959) (this is where `Mathlib.Data.List.ofFn` is imported directly) and successful run at [here](https://github.com/leanprover-community/mathlib4/actions/runs/5114799234/jobs/9195413468), `Mathlib.Data.List.Indexes` is imported so that`ofFn` is only indirectly imported

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,198 @@
+/-- the underlying function of a relation series -/
+toFun : Fin (length + 1) → α
+/-- adjacent elements are related by the said relation -/
+step : ∀ (i : Fin length), r (toFun <| Fin.castSucc i) <| toFun <| i.succ
```
```suggestion
step : ∀ (i : Fin length), r (toFun (Fin.castSucc i)) (toFun i.succ)
```
This is definitionally the same but I think it's more readable?

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,198 @@
+    | hs j ihj =>
+      by_cases H : 0 < Fin.castSucc j
+      . exact IsTrans.trans _ _ _ (ihj H) (x.step _)
+      . convert x.step _
+        simp only [not_lt, Fin.le_zero_iff] at H
+        exact H.symm
```
```suggestion
      . simp only [not_lt, Fin.le_zero_iff] at H
        rw [← H]
        exact x.step _
```
I've been a bit edgy about non-terminal `convert`s ever since I had a discussion with Gabriel about them. This avoids the non-terminal `convert` without being any longer.

kbuzzard commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,198 @@
+      . change (i + 1 : ℕ) < (j + 1 : ℕ) at h
+        rw [Nat.lt_succ_iff, le_iff_lt_or_eq] at h
+        rcases h with (h|h)
+        . right
+          exact h
```
```suggestion
        . exact Or.inr h
```

kbuzzard reviewed (approved):
This still looks fine to me. I've made some more minor comments.

YaelDillies commented on Mathlib/Order/KrullDimension.lean:
```diff
@@ -32,67 +29,13 @@ sum of `-∞` and the Krull dimension of any other varieties.
-series of `α`. If there is no strict series `a₀ < a₁ < ... < aₙ` in `α`, then its Krull dimension
-is defined to be negative infinity; if the length of `a₀ < a₁ < ... < aₙ` is unbounded, its Krull
-dimension is defined to be positive infinity.
+Krull dimension of a preordered set `α` is the supremum of the right most index of all relation
```
```suggestion
Krull dimension of a preorder `α` is the supremum of the rightmost index of all relation
```

YaelDillies commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,198 @@
+instance : CoeFun (RelSeries r) (fun x ↦ Fin (x.length + 1) → α) :=
+{ coe := RelSeries.toFun }
+
+instance : Preorder (RelSeries r) :=
+  Preorder.lift fun x => x.length
```
That looks like a very poor notion of "less than". I would have rather expected `a ≤ b ↔ a.toList.sublist b.toList`. Do you have a reason behind this?

jjaassoonn replied:
This was originally for krull dimensions of a ring. So only the length of a chain is important.

jjaassoonn replied:
When considering `\covby` relation of submodules, this weak preorder will give the length of a module

YaelDillies replied:
How will it "give" it?

jjaassoonn replied:
With this preorder we can also define type class of finite dimensional rings to be order with a top; infinite dimensional rings to be nonempty and without top.

YaelDillies replied:
That sounds unnatural. I would expect you to use `no_max_order` instead of `no_top_order` (for the second one).

YaelDillies commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,198 @@
+
+/-- Relation series of `r` and nonempty list of `α` satisfying `r`-chain condition bijectively
+corresponds to each other.-/
+def Equiv : RelSeries r ≃ {x : List α | x ≠ ∅ ∧ x.Chain' r} where
```
```suggestion
protected def Equiv : RelSeries r ≃ {x : List α | x ≠ ∅ ∧ x.Chain' r} where
```
else the `≃` notation will break once you `open RelSeries`.

YaelDillies commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,198 @@
+  . rw [List.get_ofFn]
+    congr 1
+
+lemma toList_not_empty (x : RelSeries r) : x.toList ≠ ∅ := fun m =>
```
```suggestion
lemma toList_ne_empty (x : RelSeries r) : x.toList ≠ ∅ := fun m =>
```

YaelDillies commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,197 @@
+`a_0, a_1, ..., a_n` such that `r a_i a_{i+1}` for all `i < n`
+-/
+structure RelSeries where
+/-- the number of inequalities in the series -/
```
```suggestion
/-- The number of inequalities in the series -/
```

YaelDillies commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,197 @@
+structure RelSeries where
+/-- the number of inequalities in the series -/
+length : ℕ
+/-- the underlying function of a relation series -/
```
```suggestion
/-- The underlying function of a relation series -/
```

YaelDillies commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,197 @@
+length : ℕ
+/-- the underlying function of a relation series -/
+toFun : Fin (length + 1) → α
+/-- adjacent elements are related by the said relation -/
```
```suggestion
/-- Adjacent elements are related -/
```

YaelDillies commented on Mathlib/Order/KrullDimension.lean:
```diff
@@ -32,67 +29,13 @@ sum of `-∞` and the Krull dimension of any other varieties.
+unbounded, its Krull dimension is defined to be positive infinity.
 -/
 noncomputable def krullDim : WithBot (WithTop ℕ) :=
-  ⨆ (p : StrictSeries α), p.length
+  ⨆ (p : LTSeries α), p.length
```
This is basically `Order.pred ((Set.univ : Set α) .chain_height : WithBot ℕ∞)` (once generalised to an arbitrary relation).

eric-wieser replied:
I think it's fundamentally not a pred_order

jjaassoonn replied:
@YaelDillies do you think changing `s.chainHeight` to be `withBot ℕ∞` a good idea?

alreadydone replied:
Discussions on Zulip seems to conclude it's nicer to use `Fin` indexed tuples rather than `List`, and `RelSeries` rather than `List.chain'`. I think one reason it's nicer is that we restrict to positive length tuples and have direct access to the first and last terms of the chain without worrying about the `List.nil` case. Of course we may bundle the `≠ List.nil` condition or use a pair `(a,l)` as in `List.chain` to form the `RelSeries` type, but why not just generalize `CompositionSeries` given that there's already API around it. For example, `snoc` in #3852 and `append` in #3858 are already there, and now that we've `StrictMono` with the `step` condition we can simplify the proofs by copying from `CompositionSeries`. Worth noting that `append` in #3858 has length `p.length + q.length + 1` rather than `p.length + q.length` as in `CompositionSeries`. I think the latter is usually what we want. (For `List` there are List.Chain'.append and [List.Chain'.append_overlap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/List/Chain.html#List.Chain'.append_overlap).)

Using `RelSeries` instead of `List.chain'` would probably mean abandonment of the `Set.chainHeight` API, which is unfortunate, and I wonder whether @erdOne has anything to say about it. `Set.chainHeight` is "off by one" and It's probably not worth redefining it if we're not gonna use it, but we may use an OrderIso between the two to transfer results if we want.

Now, some comments about the name `krullDim`: since you'd like to use it to define length of modules, I'd suggest calling it some generic name like `supSeriesLength` or `supChainLength`, and reserving `krullDim` for `CommRing.krullDim` or `TopologicalSpace.krullDim`. You won't want it to clash with [Krull dimension of modules](https://en.wikipedia.org/wiki/Deviation_of_a_poset) which is distinct from the length and defined in a totally different way.

alreadydone replied:
Oh, I just noticed #5881 already generalized `chainHeight` to arbitrary relations. Maybe we should just define `pred: ℕ∞ ≃o ℕ±∞`, prove that `pred set.univ.chainHeight = krullDim α` and `pred s.chainHeight = krullDim ↥s`, and start transferring result from `chainHeight` to `krullDim` ...

jjaassoonn replied:
> Now, some comments about the name krullDim: since you'd like to use it to define length of modules, I'd suggest calling it some generic name like supSeriesLength or supChainLength, and reserving krullDim for CommRing.krullDim or TopologicalSpace.krullDim. You won't want it to clash with [Krull dimension of modules](https://en.wikipedia.org/wiki/Deviation_of_a_poset) which is distinct from the length and defined in a totally different way.

Thanks for pointing this out

alreadydone replied:
> in #6277, I made the other way round, I rewrite `JordanHolder.lean` completely in terms of `RelSeries`

Yeah, that's exactly what I mean and I think #6277 is in the right direction. Fantastic work!

YaelDillies reviewed (commented):
This seems to duplicate a lot of material from `Order.Height`. Could you try unifying stuff here rather than adding yet another structure?

alreadydone reviewed (commented):
Here are my preliminary comments about the general design; thanks to @jcommelin for alerting me of this open PR.

jjaassoonn commented:
> I doubt that you really want to use \covby / CompositionSeries in the definition of length of modules. 
In #6277 I have rewritten `JordanHolder.lean` in terms of `RelSeries` 
> There are three ways of demonstrably equivalent definitions of length, and I prefer the first one because you can't use krullDim for the other two:
> 1. the supremum of lengths of all RelSeries of <, i.e. krullDim for the relation <.
> 2. the supremum of lengths of all RelSeries of < with the first term bot and the last term top; this is equivalent to 1 because any RelSeries of < can be extended to make the first term bot and the last term top.
> 3. the length of any RelSeries of \covby with the first term bot and the last term top if any such series exists (by Jordan-Holder they all have the same length), and infinity if none exists.

I have proved in a Lean 3 project that 1 <-> 3 at [here](https://github.com/FMLJohn/dimension_theory/blob/622b4c4125ed45a9f28b5317c5395018aa5bc59f/src/module_length.lean#L63) and I am trying to translate the project in #6277

jjaassoonn commented:
> > That looks like a very poor notion of "less than". I would have rather expected `a ≤ b ↔ a.toList.sublist b.toList`. 

> Do you have a reason behind this?

My only defence is that for this `RelSeries`, I only care about length, and I argue that since `RelSeries` is a separate type instead of being e.g. subtype of lists, so there is no harm for defining the order to be one induced by length

jjaassoonn commented:
> > my point was that this definition makes infinite dimensional rings something already in mathlib. There is a trichotomy of zero type, finite dimensional and infinite dimensional type, they corresponds to existing type class of Empty, TopOrder and NoTopOrder on LTSeries.
> 
> That seems to be abusing the typeclass system and doesn't really justify introducing these two instances: as I said you could use `TopOrder (Set.range (LTSeries α).length)`, and for the empty case simply `IsEmpty α`. Also, I think you don't always want to use trichotomy since many results hold both for empty types and finite-dimensional types, which you can't state using a disjunction of typeclasses; so I expect it's more convenient to have a `FiniteDimensional` typeclass that says `krullDim < ⊤` and add in a `Nonempty` condition when necessary. (For example, quotients and localizations of `FiniteDimensional` rings are `FiniteDimensional`.)
> 
> That said, I have no idea whether it's better to reuse existing typeclasses or introducing new ones from a performance perspective in general.

I have removed the awkward order on `RelSeries` in #6277, instead I introduced some new type classes such as [`Rel.FiniteDimensional`](https://github.com/leanprover-community/mathlib4/blob/1e68c843a84aedc0f0d12397c35e946adb25976f/Mathlib/Order/RelSeries.lean#L77), [`FiniteDimensionalType`](https://github.com/leanprover-community/mathlib4/blob/1e68c843a84aedc0f0d12397c35e946adb25976f/Mathlib/Order/RelSeries.lean#L669), [`FiniteLengthModule`](https://github.com/leanprover-community/mathlib4/blob/1e68c843a84aedc0f0d12397c35e946adb25976f/Mathlib/Algebra/Module/Length.lean#L33)

jjaassoonn commented:
I have now removed the weird ordering by length on `RelSeries`, hopefully now this PR is less controversial

alreadydone commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,261 @@
+/-- a relation `r` is said to be finite dimensional iff there is a relation series of `r` with the
+  maximum length. -/
+class FiniteDimensional where
+  /-- the longest relation series-/
+  longestRelSeries : RelSeries r
+  /-- the longest relation series is longer than any other series -/
+  is_longest : ∀ (y : RelSeries r), y.length ≤ longestRelSeries.length
```
Is there a reason that you want to keep track of a particular longest RelSeries? Notice that data held by a class is supposed to be canonical structure on a type rather than an arbitrarily chosen distinguished RelSeries. Personally I'd just define this class to consists of a single field: `exists_longest : ∃ longest : RelSeries r, ∀ y : RelSeries r, y.length ≤ longest.length`. This way the class would be Prop-valued and therefore couldn't be involved in any diamonds, and also conform to the design of [Finite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Finite/Defs.html#Finite), [Module.Finite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Finiteness.html#Module.Finite) / [FiniteDimensional](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/FiniteDimensional.html#FiniteDimensional), [Algebra.FiniteType](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/FiniteType.html#Algebra.FiniteType), etc.

jjaassoonn replied:
I was trying to avoid adding non-computable, but I guess I don't really care about computability issues. So I will change it to a `Prop` valued one

jjaassoonn replied:
changed

alreadydone commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,261 @@
+/-- a relation `r` is said to be infinite dimensional iff there exists relation series of arbitrary
+  length. -/
+class InfiniteDimensional where
+  /-- there should a relation series with length at least `n` for all natural number `n` -/
+  longRelSeries : ℕ → RelSeries r
+  /-- there should a relation series with length at least `n` for all natural number `n` -/
+  longRelSeries_length : ∀ (n : ℕ), n ≤ (longRelSeries n).length
```
Similarly, I would define this class to consists of a single field `exists_length_ge : ∀ n : ℕ, ∃ s : RelSeries r, n ≤ s.length`.

jjaassoonn replied:
changed to this version

alreadydone commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,261 @@
+end Rel
+
+/-- A type is finite dimensional if its `LTSeries` has bounded length. -/
+class FiniteDimensionalType (γ : Type _) [Preorder γ] extends
+  Rel.FiniteDimensional ((. < .) : γ → γ → Prop)
```
I think
```suggestion
abbrev FiniteDimensionalType (γ : Type _) [Preorder γ] :=
  Rel.FiniteDimensional ((. < .) : γ → γ → Prop)
```
would be reasonable. (Same below)

jjaassoonn replied:
changed

alreadydone reviewed (commented):
Thanks for removing the weird order, but there are still some issues with the design of the classes.

jjaassoonn commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,261 @@
+
+variable {β : Type _} [PartialOrder β]
+
+lemma Monotone (x : LTSeries β) : Monotone x :=
```
changed and removed unnecessary assumption `PartialOrder`, so deleted the section

jjaassoonn commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,261 @@
+  · rw [List.get_ofFn]
+    congr 1
+
+lemma toList_ne_empty (x : RelSeries r) : x.toList ≠ ∅ := fun m =>
```
changed

jjaassoonn commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,261 @@
+
+/-- a relation series with length at least `n` if the relation is infinite dimensional -/
+def relSeriesOfInfiniteDimensional [InfiniteDimensional r] (n : ℕ) : RelSeries r :=
+  (InfiniteDimensional.longRelSeries n)
```
changed

YaelDillies commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -0,0 +1,255 @@
+
+variable (α) [Preorder α]
+/--
+If `α` is a preordered set, a series ordered by less than is a relation series of the less than
```
```suggestion
If `α` is a preorder, a series ordered by less than is a relation series of the less than
```

alreadydone reviewed (approved):
LGTM; pushed [some adjustments](https://github.com/leanprover-community/mathlib4/pull/3878/files/4169caf0dfc3a419adc54c89291516aecf4021f5..faadb7d5b926ee2270f9e520939294df53eb713b).

alreadydone commented on Mathlib/Order/RelSeries.lean:
```diff
@@ -129,71 +98,61 @@ lemma toList_ne_empty (x : RelSeries r) : x.toList ≠ [] := fun m =>
-noncomputable def inhabited_of_infiniteDimensional [r.InfiniteDimensional] : Inhabited α :=
-  ⟨(RelSeries.withLength r 0) 0⟩
+/-- If a relation on `α` is infinite dimensional, then `α` is nonempty. -/
+lemma nonempty_of_infiniteDimensional [r.InfiniteDimensional] : Nonempty α :=
```
I think it makes little sense to make a def to produce an arbitrary default element, so I changed it to Nonempty.

alreadydone commented:
maintainer merge

jjaassoonn commented:
Did I do something stupid?

jjaassoonn commented:
i think I did something stupid with git @alreadydone

eric-wieser commented:
I think you'll need to open a new PR from the same branch to recover here.
